channeldb/invoices.go

```diff
@@ -1064,33 +1062,29 @@ func deserializeInvoice(r io.Reader) (Invoice, error) {
 }
 
 func acceptOrSettleInvoice(invoices, settleIndex *bbolt.Bucket,
-	invoiceNum []byte, amtPaid lnwire.MilliSatoshi,
-	updateInvoice InvoiceUpdateCallback) (*Invoice, error) {
+	invoiceNum []byte, updateInvoice InvoiceUpdateCallback) (
+	*Invoice, error) {
 
 	invoice, err := fetchInvoice(invoiceNum, invoices)
 	if err != nil {
 		return nil, err
 	}
 
-	// If the invoice is still open, check the htlc parameters.
+	preUpdateState := invoice.Terms.State
```

Comment:
I don't see how there can be races. The invoice struct is populated via `fetchInvoice` in this function. The pointer isn't known anywhere else.  Everything is executed synchronously.

>How about the method return the new state or updates, rather than mutating the invoice?

See other comment about the need to create an object that describes the required changes. It is possible, but I don't know if it is better.

Reply:
to me the single callback is easier to reason about, and helps keep all of the logic in one place. what is the concern about modifying the invoice? it's not clear to me why fetch -> modify -> write would be an issue

Reply:
Fetch -> modify -> write is still in place, but with a `InvoiceUpdateDesc` object to constrain the mutations that can take place.

Reply:
Yeah I eventually also came around to the single call back direction once I viewed the change as a whole.